src/types/common.d.ts

```diff
@@ -0,0 +1,392 @@
+  x: number;
+  h: number;
+  y: number
+}
+
+export interface IText {
+  options: any;
+  content: string;
+}
+
+export interface IStaveLineRenderOptions {
+  draw_start_arrow: boolean;
+  padding_left: number;
+  text_justification: number;
+  color: null;
+  line_width: number;
+  line_dash: null;
+  rounded_end: boolean;
+  arrowhead_length: number;
+  text_position_vertical: number;
+  draw_end_arrow: boolean;
+  arrowhead_angle: number;
+  padding_right: number
+}
+
+export interface ITextBracketRenderOptions {
+  dashed: boolean;
+  color: string;
+  line_width: number;
+  underline_superscript: boolean;
+  show_bracket: boolean;
+  dash: number[];
+  bracket_height: number
+}
+
+export interface ISpace {
+  mean: number;
+  deviation: number;
+  used: number
+}
+
+export interface IBarnoteMetrics {
+  widths: Record<string, number>
+}
+
+export interface IBendRenderOptions {
+  line_width: number;
+  release_width: number;
+  bend_width: number;
+  line_style: string
+}
+
+export interface ISize {
+  width: number;
+  height: number
+}
+
+export interface IPedalMarkingGlyph {
+  code: string;
+  y_shift: number;
+  x_shift: number;
+}
+
+export interface IGraceNoteGroupRenderOptions {
+  slur_y_shift: number
+}
+
+export interface IVibratoRenderOptions {
+  wave_height: number;
+  wave_girth: number;
+  vibrato_width: number;
+  harsh: boolean;
+  wave_width: number
+}
+
+export interface ITremoloRenderOptions {
+  stroke_spacing: number;
+  stroke_px: number;
+  font_scale: any;
+}
+
+export interface IContextGaps {
+  total: number;
+  gaps: any[]
+}
+
+export interface IKeyParts {
+  root: string;
+  type: string;
+  accidental: string
+}
+
+export interface IModifierContextState {
+  right_shift: number;
+  left_shift: number;
+  text_line: number;
+  top_text_line: number
+}
+
+export interface IRaphaelAttributes {
+  scale: string;
+  transform: string;
+  "stroke-width": number;
+  fill: string;
+  stroke: string;
+  font: string
+}
+
+export interface IRaphaelState {
+  font_weight: number;
+  font_size: number;
+  scale: ICoordinates;
+  font_family: string
+}
+
+export interface ILine {
+  column: number;
+  line: number;
+  flatLine: boolean;
+  dblSharpLine: boolean;
+  numAcc: number;
+  width: number;
+}
+
+export interface IMusicNoteValue {
+  root_index: number;
+  int_val: number;
+}
+
+export interface ITempo {
+  bpm: number;
+  dots: number;
+  duration: string;
+  name: string;
+}
+
+export interface IPhrase {
+  x: number;
+  type: number;
+  text: string;
+  width: number;
+  draw_width: number;
+}
+
+export interface ISymbolBlock {
+  vAlign: boolean;
+  symbolModifier: number;
+  text: string;
+  yShift: number;
+  xShift: number;
+  width: number;
+  glyph: Glyph;
+  symbolType: number;
+}
+
+export interface IStyle {
+  lineWidth: number;
+  strokeStyle: string;
+  fillStyle: string;
+  shadowBlur: number;
+  shadowColor: string;
+}
+
+export type DrawContext = SVGContext | CanvasContext | RaphaelContext;
```

Comment:
With DrawContext being a union type, it might be difficult to keep the relevant APIs of the three contexts in sync in the future. Perhaps an interface which holds the relevant methods/fields would be better here?

Reply:
I agree, but refactoring this within the PR will make it even harder to review. I would suggest we create an issue and fix it after we merge this PR. What do you think?

Reply:
Sure. Totally fine with me.

Reply:
I'm working on migrating tests to typescript as well, and this already causes a problem there. And this is not that big change. So this will probably be fixed with next commit, which covers tests migration.